lib/media/media_source_engine.js

```diff
@@ -1513,6 +1554,181 @@ shaka.media.MediaSourceEngine = class {
+      this.sourceBuffers_[contentType].changeType(type);
+      this.sourceBufferTypes_[contentType] = mimeType;
+    } else {
+      shaka.log.debug('Change Type not supported');
+    }
+
+    // Fake an 'updateend' event to resolve the operation.
+    this.onUpdateEnd_(contentType);
+  }
+
+  /**
+   * Enqueue an operation to prepare the SourceBuffer to parse a potentially new
+   * type or codec.
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @param {string} mimeType
+   * @param {?shaka.extern.Transmuxer} transmuxer
+   * @return {!Promise}
+   */
+  changeType(contentType, mimeType, transmuxer) {
+    return this.enqueueOperation_(
+        contentType,
+        () => this.change_(contentType, mimeType, transmuxer));
+  }
+
+  /**
+   * Resets the MediaSource and re-adds source buffers due to codec mismatch
+   *
+   * @param {!Map.<shaka.util.ManifestParserUtils.ContentType,
+   *               shaka.extern.Stream>} streamsByType
+   * @private
+   */
+  async reset_(streamsByType) {
+    this.reloadingMediaSource_ = true;
+    const currentTime = this.video_.currentTime;
+    try {
+      this.eventManager_.removeAll();
+      this.sourceBuffers_ = {};
+
+      const cleanup = [];
+      for (const contentType in this.transmuxers_) {
+        cleanup.push(this.transmuxers_[contentType].destroy());
+      }
+      await Promise.all(cleanup);
+      this.transmuxers_ = {};
+
+      const previousDuration = this.mediaSource_.duration;
+      this.mediaSourceOpen_ = new shaka.util.PublicPromise();
+      this.mediaSource_ = this.createMediaSource(this.mediaSourceOpen_);
+      await this.mediaSourceOpen_;
+      this.mediaSource_.duration = previousDuration;
+
+      const sourceBufferAdded = new shaka.util.PublicPromise();
+      const sourceBuffers =
+        /** @type {EventTarget} */(this.mediaSource_.sourceBuffers);
+
+      const totalOfBuffers = streamsByType.size;
+      let numberOfSourceBufferAdded = 0;
+      this.eventManager_.listen(sourceBuffers, 'addsourcebuffer', (event) => {
+        numberOfSourceBufferAdded++;
+        if (numberOfSourceBufferAdded === totalOfBuffers) {
+          sourceBufferAdded.resolve();
+        }
+      });
+
+      for (const contentType of streamsByType.keys()) {
+        const stream = streamsByType.get(contentType);
+        this.initSourceBuffer_(contentType, stream);
+      }
+
+      // Fake a seek to catchup the playhead.
+      this.video_.currentTime = currentTime;
+
+      await sourceBufferAdded;
+    } finally {
+      this.reloadingMediaSource_ = false;
+    }
+  }
+
+  /**
+   * Resets the Media Source
+   * @param {!Map.<shaka.util.ManifestParserUtils.ContentType,
+   *               shaka.extern.Stream>} streamsByType
+   * @return {!Promise}
+   */
+  reset(streamsByType) {
+    return this.enqueueBlockingOperation_(
+        () => this.reset_(streamsByType));
+  }
+
+  /**
+   * Codec switch if necessary
```

Comment:
I would specify that this will not resolve until the codec switch is over. Like, it doesn't just queue a codec switch to happen at a later time.

Reply:
Added a better comment